### PR TITLE
Move TestOAuthExpiration to e2e

### PR DESCRIPTION
**Pull request by sanchezl**
Part of Issue https://github.com/openshift/origin/issues/23258

### CODE REVIEW

**Comment by sanchezl** [comment]
/assign @sallyom 
/assign @stlaz 
PTAL

File: test/extended/util/oauthserver/oauthserver.go [hunk 1]
```diff
@@ -59,10 +59,12 @@ var (
 	volumesDefaultMode int32 = 420
 )
 
+type NewRequestTokenOptionsFunc func(string, string) *tokencmd.RequestTokenOptions
```
Comment (enj) [inline test/extended/util/oauthserver/oauthserver.go]:
```go
type NewRequestTokenOptionsFunc func(username, password string) *tokencmd.RequestTokenOptions
```

File: test/extended/util/oauthserver/oauthserver.go [hunk 2]
```diff
@@ -97,7 +99,9 @@ func DeployOAuthServer(oc *exutil.CLI, idps []osinv1.IdentityProvider, configMap
 	if _, err := oc.AdminKubeClient().RbacV1().ClusterRoleBindings().Create(oauthSARolebinding); err != nil {
 		return nil, cleanups, err
 	}
-
+	if err := exutil.WaitForUserBeAuthorized(oc, "system:serviceaccount:"+oc.Namespace()+":e2e-oauth", "get", "oauthclients.oauth.openshift.io"); err != nil {
```
Comment (enj) [inline test/extended/util/oauthserver/oauthserver.go]: Wait for the route `/healthz` to report `200` as well.

File: test/extended/oauth/htpasswd.go [hunk 3]
```diff
@@ -53,18 +56,20 @@ var _ = g.Describe("[Suite:openshift/oauth/htpasswd] HTPasswd IDP", func() {
...
-		tokenReqOpts, cleanup, err := utiloauth.DeployOAuthServer(oc, []osinv1.IdentityProvider{htpasswdConfig}, nil, secrets)
+		newTokenReqOpts, cleanup, err := utiloauth.DeployOAuthServer(oc, []osinv1.IdentityProvider{htpasswdConfig}, nil, secrets)
 		defer cleanup()
 		o.Expect(err).ToNot(o.HaveOccurred())
+		tokenReqOpts := newTokenReqOpts("testuser", "password")
 		e2e.Logf("got the OAuth server address: %s", tokenReqOpts.Issuer)
-
-		token, err := utiloauth.RequestTokenForUser(tokenReqOpts, "testuser", "password")
+		var token string
+		o.Eventually(func() error {
```
Comment (enj) [inline test/extended/oauth/htpasswd.go]: You should not need any retries once the OAuth server is fixed up.

File: test/extended/oauth/htpasswd.go [hunk 4]
```diff
@@ -53,18 +56,20 @@ var _ = g.Describe("[Suite:openshift/oauth/htpasswd] HTPasswd IDP", func() {
...
+			token, err = tokenReqOpts.RequestToken()
```
Comment (enj) [inline test/extended/oauth/htpasswd.go]: If you did actually need a retry, the correct approach is:

```go
token, err = newTokenReqOpts("testuser", "password").RequestToken()
```

File: test/extended/oauth/expiration.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+var _ = g.Describe("[Feature:OAuthServer] [Token Expiration]", func() {
+	var oc = util.NewCLI("oauth-expiration", util.KubeConfigPath())
+	var newRequestTokenOptions oauthserver.NewRequestTokenOptionsFunc
+	var oauthServerCleanup func()
+
+	g.BeforeEach(func() {
+		var err error
+		newRequestTokenOptions, oauthServerCleanup, err = deployOAuthServer(oc)
+		o.Expect(err).ToNot(o.HaveOccurred())
+		<-time.After(5 * time.Second) // TODO give RBAC, pod, & routes time to settle
```
Comment (enj) [inline test/extended/oauth/expiration.go]: Fix

File: test/extended/oauth/expiration.go [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+			})
+			o.Expect(err).ToNot(o.HaveOccurred())
+		})
+
+		g.AfterEach(func() {
+			oc.AdminOAuthClient().OauthV1().OAuthClients().Delete(oAuthClientResource.Name, &metav1.DeleteOptions{})
+		})
+
+		g.Context("to generate tokens that do not expire", func() {
+
```
Comment (enj) [inline test/extended/oauth/expiration.go]: IMO `accessTokenMaxAgeSeconds := 0` right here is easier to read.

File: test/extended/util/oauthserver/oauthserver.go [hunk 7]
```diff
@@ -59,10 +59,12 @@ var (
 	volumesDefaultMode int32 = 420
 )
 
+type NewRequestTokenOptionsFunc func(string, string) *tokencmd.RequestTokenOptions
+
```
Comment (stlaz) [inline test/extended/util/oauthserver/oauthserver.go]: nit: could you please split the changes done to the util oauth server into a separate commit?

Comment (sanchezl) [reply test/extended/util/oauthserver/oauthserver.go]: Split into separate PR: https://github.com/openshift/origin/pull/23528

File: test/extended/oauth/expiration.go (hunk 5 above)
Comment (sanchezl) [reply test/extended/oauth/expiration.go]: Done

File: test/extended/oauth/htpasswd.go (hunk 4 above)
Comment (sanchezl) [reply test/extended/oauth/htpasswd.go]: Removed retry logic.

File: test/extended/oauth/htpasswd.go (hunk 3 above)
Comment (sanchezl) [reply test/extended/oauth/htpasswd.go]: Removed retries.

File: test/extended/util/oauthserver/oauthserver.go (hunk 2 above)
Comment (sanchezl) [reply test/extended/util/oauthserver/oauthserver.go]: Done in https://github.com/openshift/origin/pull/23528

File: test/extended/util/oauthserver/oauthserver.go (hunk 1 above)
Comment (sanchezl) [reply test/extended/util/oauthserver/oauthserver.go]: Done

**Comment by sallyom** [comment]
/lgtm

will rebase and keep testing [remove-flaky-htpasswd PR](https://github.com/openshift/origin/pull/23421) once this merges

**Pull request merged**